docs(abac): add initial ABAC Condition documentation (#583)
openfga/openfga.dev: docs/content/modeling/abac-conditions.mdx

        "expression":"current_time <= grant_time + grant_duration",
        "parameters": {
          "current_time": {
              "type_name":"TYPE_NAME_TIMESTAMP"
          },
          "grant_duration": {
              "type_name":"TYPE_NAME_DURATION"
          },
          "grant_time": {
              "type_name":"TYPE_NAME_TIMESTAMP"
          }
        }
      }
    }
  }}
  allowedLanguages={[
    SupportedLanguage.JS_SDK,
    SupportedLanguage.GO_SDK,
    SupportedLanguage.DOTNET_SDK,
    SupportedLanguage.PYTHON_SDK,
    SupportedLanguage.CLI,
    SupportedLanguage.CURL,
  ]}
/>

## Writing Conditional Relationship Tuples
Using the model above, when we [Write](https://openfga.dev/api/service#/Relationship%20Tuples/Write) relationship tuples to the OpenFGA store, then any `document#viewer` relationship with `user` objects must be accompanied by the condition `temporal_grant` because the type restriction requires it.

For example, we can give `user:anne` viewer access to `document:1` for 10 minutes by writing the following relationship tuple:

<WriteRequestViewer
  relationshipTuples={[
    {
      user: 'user:anne',
      relation: 'viewer',
      object: 'document:1',

>       condition: {

Author: todo: WriteRequestViewer needs to wired up to support conditions.

        name: 'temporal_grant',
        context: {
          grant_time: '2023-01-01T00:00:00Z',
          grant_duration: '10m'
        }
      }
    }
  ]}
  skipSetup={true}
  allowedLanguages={[
    SupportedLanguage.JS_SDK,
    SupportedLanguage.GO_SDK,
    SupportedLanguage.DOTNET_SDK,
    SupportedLanguage.PYTHON_SDK,
    SupportedLanguage.CLI,
    SupportedLanguage.CURL,
  ]}
/>

## Queries with Condition Context
Now that we have written a [Conditional Relationship Tuple](../concepts#what-is-a-conditional-relationship-tuple), we can query OpenFGA using the [Check API](https://openfga.dev/api/service#/Relationship%20Queries/Check) to see if `user:anne` has viewer access to `document:1` under certain conditions/context. That is, `user:anne` should only have access is the current timestamp is less than or equal to the grant timestamp (e.g. the time which the tuple was written) plus the duration of the grant (10 minutes). If the current timestamp is less than or equal, then you'll get a permit decision. For example,

> <CheckRequestViewer

Author: todo: CheckRequestViewer needs to be wired up to support `context`. It does not currently have support at this time.

  user={'user:anne'}
  relation={'viewer'}
  object={'document:1'}
  allowed={true}
  skipSetup={true}
  allowedLanguages={[
    SupportedLanguage.JS_SDK,
    SupportedLanguage.GO_SDK,
    SupportedLanguage.DOTNET_SDK,
    SupportedLanguage.PYTHON_SDK,
    SupportedLanguage.CLI,
    SupportedLanguage.CURL,
  ]}
/>

but if the current time is outside the grant window then you get a deny decision. For example,

<CheckRequestViewer
  user={'user:anne'}
  relation={'viewer'}
  object={'document:1'}
  allowed={false}
  skipSetup={true}
  allowedLanguages={[
    SupportedLanguage.JS_SDK,
    SupportedLanguage.GO_SDK,
    SupportedLanguage.DOTNET_SDK,
    SupportedLanguage.PYTHON_SDK,
    SupportedLanguage.CLI,
    SupportedLanguage.CURL,
  ]}
/>

Similarly, we can use the [ListObjects API](https://openfga.dev/api/service#/Relationship%20Queries/ListObjects) to return all of the documents that `user:anne` has viewer access to given the current time. For example,

> <ListObjectsRequestViewer

Author: todo: wire up `context` with the ListObjectsRequestViewer.

---
sidebar_position: 8
slug: /modeling/abac-conditions
description: Modeling relationships with ABAC Conditions
---

import {
  AuthzModelSnippetViewer,
  WriteRequestViewer,
  CheckRequestViewer,
  ListObjectsRequestViewer,
  languageLabelMap,
  SdkSetupPrerequisite,
  SupportedLanguage,
  WriteAuthzModelViewer,
} from '@components/Docs';

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # ABAC Conditions

Author: Ok, that will overload "Conditional Relationship Tuples" with "Conditional Relationships" but if that's what you prefer then I'll ship it.

## Overview
ABAC Conditions allow you to model more complex authorization modeling scenarios involving Attribute-based Access Control (ABAC) policies. Conditions are used in [Conditional Relationship Tuples](../concepts#what-is-a-conditional-relationship-tuple)

There are various use cases where ABAC Conditions can be helpful. These include, but are not limited to:

> * **Temporal Access Policies** - manage user access for a window of time.

Author: sure

> * **Temporal Access Policies** - manage user access for a window of time.

Author: added

* **IP Allowlists or Geo-fencing Policies** - limit or grant access based on an IP Address range or corporate network policy.
* **Usage-based/Feature-based Policies (Entitlements)** - enforce quoata or usage of some resource or feature.
* **Resource-attribute Policies** - define policies to access resources based on attributes/fields of the resource(s).

For more information and background context on why we added this feature, please see our blog post on [Conditional Relationship Tuples for OpenFGA](https://openfga.dev/blog/conditional-tuples-announcement).

## Defining ABAC Conditions in Models
For this example we'll use the following authorization model to demonstrate a temporal based access policy. Namely, a user can view a document if and only if they have been granted the viewer relationship AND their temporal grant policy is met.

```dsl.openfga
model
  schema 1.1

type user

type document
  relations
    define viewer: [user with temporal_grant]

> condition temporal_grant(grant_duration: duration, current_time: timestamp, grant_time: timestamp) {

Author: Updated

---
sidebar_position: 8
slug: /modeling/abac-conditions
description: Modeling relationships with ABAC Conditions
---

import {
  AuthzModelSnippetViewer,
  WriteRequestViewer,
  CheckRequestViewer,
  ListObjectsRequestViewer,
  languageLabelMap,
  SdkSetupPrerequisite,
  SupportedLanguage,
  WriteAuthzModelViewer,
} from '@components/Docs';

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # ABAC Conditions

Reviewer: ```suggestion
# Conditions
```

## Overview

> ABAC Conditions allow you to model more complex authorization modeling scenarios involving Attribute-based Access Control (ABAC) policies. Take a look at the [Conditions](../concepts.mdx#what-is-an-abac-condition) and [Conditional Relationship Tuples](../concepts.mdx#what-is-a-conditional-relationship-tuple) concepts for a quick overview.

Reviewer: ```suggestion
Conditions allow you to model more complex authorization modeling scenarios involving attributes and can be used to represent some Attribute-based Access Control (ABAC) policies. Take a look at the [Conditions](../concepts.mdx#what-is-a-condition) and [Conditional Relationship Tuples](../concepts.mdx#what-is-a-conditional-relationship-tuple) concepts for a quick overview.
```

There are various use cases where ABAC Conditions can be helpful. These include, but are not limited to:

* [**Temporal Access Policies**](https://github.com/openfga/sample-stores/tree/main/stores/temporal-access) - manage user access for a window of time.
* [**IP Allowlists or Geo-fencing Policies**](https://github.com/openfga/sample-stores/tree/main/stores/ip-based-access) - limit or grant access based on an IP Address range or corporate network policy.
* [**Usage-based/Feature-based Policies (Entitlements)**](https://github.com/openfga/sample-stores/tree/main/stores/advanced-entitlements) - enforce quoata or usage of some resource or feature.
* [**Resource-attribute Policies**](https://github.com/openfga/sample-stores/tree/main/stores/groups-resource-attributes) - define policies to access resources based on attributes/fields of the resource(s).

For more information and background context on why we added this feature, please see our blog post on [Conditional Relationship Tuples for OpenFGA](https://openfga.dev/blog/conditional-tuples-announcement). 

> ## Defining ABAC Conditions in Models

Reviewer: ```suggestion
## Defining Conditions in Models
```

For this example we'll use the following authorization model to demonstrate a temporal based access policy. Namely, a user can view a document if and only if they have been granted the viewer relationship AND their temporal grant policy is met.

```dsl.openfga
model
  schema 1.1

type user

type document
  relations
    define viewer: [user with temporal_grant]

condition temporal_grant(grant_duration: duration, current_time: timestamp, grant_time: timestamp) {
  current_time <= grant_time + grant_duration
}
```

:::note

> The type restriction for `document#viewer` requires than any `user` subject that is written in the relationship tuple must be accompanied by the `temporal_grant` condition. This is denoted by the `user with temporal_grant` specification.

Reviewer 2: ```suggestion
The type restriction for `document#viewer` requires that any `user` subject that is written in the relationship tuple must be accompanied by the `temporal_grant` condition. This is denoted by the `user with temporal_grant` specification.
```

:::

> Now we transform the DSL model into JSON syntax and write the model.

Reviewer: No need to mention about the transformation, just say write the model

<WriteAuthzModelViewer
  authorizationModel={{
    "schema_version":"1.1",
    "type_definitions": [
      {
        "type":"user"
      },
      {
        "type":"document",
        "relations": {
          "viewer": {
            "this": {}
          }
        },
        "metadata": {
          "relations": {
            "viewer": {
              "directly_related_user_types": [
                {
                    "type":"user",
                    "condition":"temporal_grant"
                }
              ]
            }
          }
        }
      }
    ],
    "conditions": {
      "temporal_grant": {
        "name":"temporal_grant",
        "expression":"current_time <= grant_time + grant_duration",
        "parameters": {
          "current_time": {
              "type_name":"TYPE_NAME_TIMESTAMP"
          },
          "grant_duration": {
              "type_name":"TYPE_NAME_DURATION"
          },
          "grant_time": {
              "type_name":"TYPE_NAME_TIMESTAMP"
          }
        }
      }
    }
  }}
  skipSetup={true}
  allowedLanguages={[

>     SupportedLanguage.JS_SDK,

Reviewer: Can you remove .NET and Python from here for now?

    SupportedLanguage.GO_SDK,
    SupportedLanguage.CURL,
  ]}
/>

but if the current time is outside the grant window then we don't get the object in the response. For example,

<ListObjectsRequestViewer
  objectType="document"
  relation="viewer"
  user="user:anne"
  context={{current_time: "2023-01-01T00:10:01Z"}}
  expectedResults={[]}
  skipSetup={true}
  allowedLanguages={[
    SupportedLanguage.JS_SDK,
    SupportedLanguage.GO_SDK,
    SupportedLanguage.CURL,
  ]}
/>

:::note
When evaluating a condition at request time, the context written/persisted in the relationship tuple and the context provided at request time are merged together into a single evaluation context.

If you provide a context value in the request context that is also written/persisted in the relationship tuple, then the context values written in the relationship tuple take precedence. That is, the merge strategy is such that persisted context has higher precedence than request context.
:::

## Examples
For more examples take a look at our [Sample Stores](https://github.com/openfga/sample-stores) repository. There are various examples with ABAC models in that repository.

## Supported Parameter Types

> The following table enumerates the lsit of supported parameter types. The more formal list is defined in specifically in https://github.com/openfga/openfga/tree/main/internal/condition/types.

Reviewer 2: ```suggestion
The following table enumerates the list of supported parameter types. The more formal list is defined in https://github.com/openfga/openfga/tree/main/internal/condition/types.
```

| map\<T\>             | TYPE_NAME_MAP             | A map whose keys are strings and whose values are values of generic type T.<br /><br />Any map value must have string keys, only the value types can vary.                                                                                                            | map\<int\> - \{"x": -1, "y": 1\}<br />map\<string\> - \{"key": "value"\}                          |
| ipaddress          | TYPE_NAME_IPADDRESS       | A custom value type specified as a string representation of an IP Address.                                                                                                                                                                                        | "192.168.0.1"                                                                           |

## Limitations
* The size of the condition `context` parameter that can be written alongside a relationship tuple is limited to 32KB in size.

* The size of the condition `context` parameter for query requests (e.g. Check, ListObjects, etc..) is not explicitly limited, but the OpenFGA server has an overall request size limit of 512KB at this time.

> * We're still working on the changes to support ABAC Conditions in the official FGA CLI and various OpenFGA SDKs including: .Net and Python. At this moment you cannot Write conditional relationship tuples with these tools and/or query (e.g. Check, ListObjects, etc..) OpenFGA with condition context.

Reviewer 2: > OpenFGA SDKs including: .Net and Python

@rhamzeh can you confirm that there are OK?

> * We're still working on the changes to support ABAC Conditions in the official FGA CLI and various OpenFGA SDKs including: .Net and Python. At this moment you cannot Write conditional relationship tuples with these tools and/or query (e.g. Check, ListObjects, etc..) OpenFGA with condition context.

Reviewer: ```suggestion
* We're still working on the changes to support Conditions in the official FGA CLI and various OpenFGA SDKs including: .NET and Python. At this moment you cannot Write conditional relationship tuples with these tools and/or query (e.g. Check, ListObjects, etc..) OpenFGA with condition context.
```
